### ISSUE TITLE
Mount remote share as local share if on the same instance

### ISSUE THREAD

**Issue by PVince81** [owncloud/core#9090]
### Steps to reproduce
1. Setup a server "localserver" and enable the external storage app
2. Setup two users "user1" and "user2"
3. As "user1" share a folder "shared_folder" with link
4. Login as "user2"
5. Open the link from step 3
6. Click "Add to ownCloud" and enter the URL to "localserver"
7. Navigate into "shared_folder"
8. Login as "user1" and open the share dropdown
### Expected result

At step 8, "user1" must see that "shared_folder" was shared with "user2" as a local share.
"shared_folder" should become a local shared folder.
### Actual result

"shared_folder" for "user2" is a remote mount of a local folder.
### More information

OC7 git master baf46c74cda834bbdb4ecf4669587b92b56fe9a9

This is a nice to have, probably something for OC 8.

@icewind1991 @schiesbn @MTRichards @craigpg @karlitschek

**Comment by PVince81** [comment]
@MTRichards @cmonteroluque would be good to get a comment on this proposal.

Because without this it means that chains of federated shares could be created to local servers, and people might also be able to bypass permissions, see https://github.com/owncloud/core/issues/16675

**Comment by nickvergessen** [comment]
Would be nice to have a way to avoid/prevent that. Otherwise smashbox remote share testing will become a nightmare (since you need clients with different urls, etc.)

**Comment by MTRichards** [comment]
While there are use cases for mounting a local share as a remote share, it breaks the paradigm that we have.

We should be flattening a federated share with a local user into a local share, or we are essentially bypassing various security mechanisms.

**Comment by nickvergessen** [comment]
Just saying, if you are not allowed to share locally, but remotely, you can also set up a dummy oC which accepts the remote share and then remote shares it back to the old instance. So this permission combination (allowing remote without allowing local) is just a no-brainer situation which we should invest too much time into.

But don't get me wrong, detecting local shares and doing them local instead is fine, as it will increase the sync/connection performance. I'm just saying that we should care about "bypassing" the group-restriction setting for local shares, when remote shares are allowed.

**Comment by schiesbn** [comment]
The more I think about it the more I agree with @MTRichards.
I think the paradigm change is the most important one. Right now the owner of the file decides how to share a file with whom. In most cases, if I send you a link instead of a internal share i did it by purpose. Not sure if this should come with the permissions for you to create a completely independent share on my side, which will for example also bypass the expire date from the share link.

**Comment by schiesbn** [comment]
> Further with sharing 2.0 we will have different share providers for federated sharing and for user,group, link sharing. It is probably best to draw sharp lines between the features, also for maintainability and complexity reasons.

@rullzer this discussion is probably also interesting for you.

**Comment by MorrisJobke** [comment]
> In most cases people have good reasons why they send you a internal share or a link.

I for my self don't care about this that much. i just want to share with a person and not think about the technical background -> I share by link, because it's easy accessible on each platform with previews and the other person then is able to hook this into it's own owncloud-compatible service. I don't want to ask the owner "Can you please reshare this with my account abc@def.ghi because we are on the same instance and you only send a link." The owner already granted the permission the read the data. So I don't think we want to distinguish (permission-wise) between link, local or remote share.

Just my cents to this topic.

**Comment by MorrisJobke** [comment]
> But this shouldn't affect this discussion. (federated) sharing is already complex enough, don't make it even more complex by adding all possible features just because someone might want to use it.

Arguing with complexity is the wrong approach here. If it is too complex we maybe want to rethink the stuff. (see sharing 2.0 work by @rullzer) And we definitely should make the handling of all the different shares less complex, because nowadays there is always the fear that touching the share.php will cause huge trouble :see_no_evil:

**Comment by schiesbn** [comment]
> Arguing with complexity is the wrong approach here. If it is too complex we maybe want to rethink the stuff. (see sharing 2.0 work by @rullzer)

There are two types of complexity. Code complexity and the complexity of the solution in the sense of a well defined feature set, how the user interact with the features and the interaction between the components. The code complexity of the current sharing is way to high and we address this with sharing 2.0. But we should also take care that the solution as such doesn't become to complex with different ways to achieve the same, transforming one share magically to a different share, different settings etc. Looking at the second type of complexity sharing is already quite complex and I think we shouldn't increase it but rather try to decrease it at some places.

**Comment by rullzer** [comment]
@cmonteroluque I don't think adding this ticket to 9.0 is a good idea or is even feasible.

We are touching the sharing code in so many places that I really don't feel that confortable adding such a major thing. It has the potential to break things.

Federated shares => local shares. Well that could work I guess. But it would need some additional logic.

However the other use case that keeps popping up. Converting mounted link shares to local shares. I'm not sure how to do that. Since there is a big difference between link shares and "normal" shares. 

* Link shares can have expiration dates, normal shares not
* Normal shares can have reshare permissions, link shares not? Or by definition?
* Normal shares have fine grained permissions, link shares are read only, or do whatever you want

And then there is the whole UI thing of how to present this to the user? If I share with foo@bar.com and all of sudden 'John Doe' shows up in my share list. Users might get very confused.

I am not saying it can't be done. I'm saying that this is the kind of feature we really need the design week for to talk about.

**Comment by PVince81** [comment]
@rullzer regarding the conversion, it should happen before it even touches the provider. If a user uses the share dialog to share with another user, the OCS API should identify that this is in fact a local user and then use the local share provider instead of the remote share provider. (no need to have them know each other)

**Comment by PVince81** [comment]
...this, ignoring link shares. With link shares I guess it might be tricky

**Comment by rullzer** [comment]
If we ignore link shares then it is much easier.

**Comment by cmonteroluque** [comment]
@rulizer I think the confusion can be abated by documentation or even, if we want, by a dialog that informs that the federated share is being converted to a local one. We could leave link shares for a later release, couldn't we? @PVince81 @schiesbn @MorrisJobke @karlitschek @MTRichards

**Comment by PVince81** [comment]
As discussed with @rullzer, if we ignore share links for now, one acceptable approach is to have the Share API level detect whether a federated share target is actually a local user. Depending on this, if it's a local user, it will forward it to the "local shares manager" instead of the "federated shares manager"

**Comment by PVince81** [comment]
Another approach could be at an even earlier level: whenever the autocomplete delivers the federated share result, it could also return a flag that tells whether it's actually a local user or remote user.
Then the JS code could call the proper OCS API, either fed share or local share, based on that flag.

The problem with this approach is that mobile clients and desktop clients will have to implement that same logic. But with the previous solution, on Share API level, they only need to call the Share API.

According to @rullzer this needs more time to be properly thought through, especially in regards to corner cases (disabled sharing / sharing limited to groups, etc)
Also there might be an opportunity to add the sharing abstraction level described in https://github.com/owncloud/core/issues/21602

Considering this, this task would be better deferred to 9.1 to make sure it can be done right.

@cmonteroluque

**Comment by rullzer** [comment]
Basically as I have discussed with @schiesbn we need to think how to do it. Instead of hacking it in. The share providers and share manager have been designed in such a way that there is separation of concerns. Hacking it in now breaks all that.

And as stated we have a lot of weird use cases now that might or might not make this possible.

I do think #21602 and this ticket fit nicely together. But we need to properly think about how to handle this in a way that we keep the code nice, clean and maintainble and have those advantages.

I agree that 9.1 would be a better place for this.

**Comment by PVince81** [comment]
Discussed with @karlitschek and @cmonteroluque , moving to 9.1

**Comment by cmonteroluque** [comment]
agreed

**Comment by MorrisJobke** [comment]
00005427

**Comment by cdamken** [comment]
This case is related https://github.com/owncloud/enterprise/issues/1326

**Comment by PVince81** [comment]
Moving to 9.2

**Comment by PVince81** [comment]
@pmaier1 not sure if we discussed this before

**Comment by felixboehm** [comment]
@pmaier1 needs PM follow up

**Comment by pmaier1** [comment]
@PVince81 yes we did. needs effort estimation and scheduling

**Comment by PVince81** [comment]
Once we decide to go for this we'll also need a repair step / occ command to convert already existing shares to local ones.

We need to properly think this through in regards to many topics:
- security
- how to detect local share considering that there can be more than one trusted domain
- oc_filecache adjustments: need to delete the oc_storages entry and remap potential metadata to the owner's storage
    - metadata like versions, favorite, etc
- other yet to be discovered trickiness

**Comment by PVince81** [comment]
Very rough estimate, at least 3-5md for just the detection and auto-add

If we want to add a repairs step and cover the potential metadata issues it could be 1-2 weeks due to complexity and testing.
